db_sql.go

```diff
@@ -469,89 +539,155 @@ func SqlCreateDatabase(conx *DBConnection) error {
-					migrateError = tx.Commit()
-				} else {
-					tx.Rollback()
-					migrateError = err.(error)
-				}
-			}()
+	if upToDate == true {
+		conStr := fmt.Sprintf("host=%v user=%v password=%v dbname=%v sslmode=disable port=%v", conx.Host, conx.User, conx.Password, conx.Database, conx.Port)
+		_, err := migration.Open(conx.Driver, conStr, createMigrations())
+		if err != nil {
+			return false, err
+		}
+		return true, err
+	}
 
-			version, eGetVersion := readSchemaVersion(tx, schema_name)
-			if eGetVersion != nil {
-				panic(eGetVersion)
-			}
+	return false, nil
+}
 
-			if version > len(migrations) {
-				panic(errors.New(fmt.Sprintf("%v database is newer than this program (%v)", schema_name, version)))
-			}
+func createVersionTable(db *sql.DB, version int) error {
+	_, err := db.Exec(fmt.Sprintf(`
+		CREATE TABLE migration_version (
+			version INTEGER
+		);
+		INSERT INTO migration_version (version) VALUES (%v)`, version))
+	return err
+}
 
-			for ; version < len(migrations); version += 1 {
-				fmt.Printf("Migrating %v to version %v\n", schema_name, version+1)
-				if _, err := tx.Exec(migrations[version]); err != nil {
-					panic(err)
-				}
-				if _, err := tx.Exec(fmt.Sprintf("UPDATE %v_version SET version = %v", schema_name, version+1)); err != nil {
-					panic(err)
-				}
-			}
-			return nil
-		} else {
-			// Match the string 'database "foo" does not exist' to detect when the database needs to be created
-			if tryCreatingDB && strings.Index(eTxBegin.Error(), "does not exist") != -1 {
-				if eCreateDB := SqlCreateDatabase(conx); eCreateDB != nil {
-					return eCreateDB
-				} else {
-					return migrateSchemaInternal(false, conx, schema_name, migrations)
-				}
-			} else {
-				return eTxBegin
-			}
+func createMigrations() []migration.Migrator {
+	var migrations []migration.Migrator
+
+	text := []string{
+		//authgroup migrations
+		`CREATE TABLE authgroup (id SERIAL PRIMARY KEY, name VARCHAR, permlist VARCHAR);
+		CREATE UNIQUE INDEX idx_authgroup_name ON authgroup (name);`,
+
+		//authsession migrations
+		`CREATE TABLE authsession (id BIGSERIAL PRIMARY KEY, sessionkey VARCHAR, identity VARCHAR, permit VARCHAR, expires TIMESTAMP);
+		CREATE UNIQUE INDEX idx_authsession_token ON authsession (sessionkey);
+		CREATE INDEX idx_authsession_identity ON authsession (identity);
+		CREATE INDEX idx_authsession_expires  ON authsession (expires);`,
+
+		`DELETE FROM authsession;`,
+
+		//authuser migrations
+		`CREATE TABLE authuser (id BIGSERIAL PRIMARY KEY, identity VARCHAR, password VARCHAR, permit VARCHAR);
+		CREATE UNIQUE INDEX idx_authuser_identity ON authuser (identity);`,
+
+		`DROP INDEX idx_authuser_identity;
+		CREATE UNIQUE INDEX idx_authuser_identity ON authuser (LOWER(identity));`,
+
+		`ALTER TABLE authuser ADD COLUMN pwdtoken VARCHAR;`,
+
+		//Change from using identity as unique reference to user account to using userid
+		`CREATE TABLE userstore (userid BIGSERIAL PRIMARY KEY, identity VARCHAR);
```

Comment:
Since all other tables are prefixed with 'auth', I think userstore should also be called authuserstore.
I would also change the old table 'authuser' to 'authuserpwd', or something indicating that this is responsible for verification.

The comment here should be a bit clearer, such as "Change from using email address as the primary identity of a user, to a 64-bit integer, which we call UserId".
And also, the comment should mention that the above migrations were from an old system, and we're bootstrapping them into the new burntsushi migration system.

Then.. I think it's much better to drop the old tables and recreate them, vs altering them to suit their new role. Since all existing data is getting saved into _temp tables, there is no virtue in incrementally altering the tables.
The downside to altering the tables in this manner, is that it's very hard for a reader to see what the current schema is, because he needs to parse all the old migrations, as well as these changes.
Far better to just have plain old "CREATE TABLE ...." statements in there.

Another thing - shouldn't identity be called 'email' here?

Also - it's highly inadvisable to create a BIGSERIAL field, and then insert explicit values into it. As we discussed, this ends up leaving the sequence generator behind the reality of what's inside the table, and the next time you try to insert an auto-generated ID, you'll get a key violation. So don't mix and match auto-generated IDs with explicit IDs. As an example of where that's happening, see the insert into authsession: the userid field is BIGSERIAL.

Another thing - why `DROP TABLE IF EXISTS authsession_temp`? There is zero ambiguity here - authsession_temp HAS to exist, so rather just write DROP TABLE, because the IF EXISTS introduces a ton of self-doubt.